torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,330 @@
+    if the tile_size = 400. For details on preprocessing, please refer to
+    ``torchtune.models.clip._transforms.CLIPImageTransform``.
+
+    Each of these tiles is further broken down into patches by a convolution operation. For example, if
+    your patch_size = 40, then each (400, 400) tile will become a grid of 10x10 patches, and your whole image will have
+    num_tiles * n_tokens -> num_tiles * (10x10 patches + 1 CLS token) -> num_tiles * 101.
+
+    Before the transformer layers, a CLS token is added to each tile as the first token.
+
+    To help the model "see" the whole image, we use positional embeddings. If your image
+    was tile-cropped, then you need to use tile positional embeddings:
+
+    - token_pos_embedding (tiled): ``torchtune.models.clip._position_embeddings.TiledTokenPositionalEmbedding``
+    - pre_tile_pos_embed: ``torchtune.models.clip._position_embeddings.TilePositionalEmbedding``
+    - post_tile_pos_embed: ``torchtune.models.clip._position_embeddings.TilePositionalEmbedding``
+
+    Otherwise, pre and post tile_pos_embed should be None. For the tokens, all you need is a simple
+    token positional embedding:
+
+    - token_pos_embedding: ``torchtune.models.clip._position_embeddings.TokenPositionalEmbedding``
+
+    All images will be considered as a stack of tiles, even if your image was not tile-cropped. In such cases,
+    your image would be composed of a single tile.
+
+    Args:
+        patch_grid_size (int): The size of a squared grid that represents how many
+             patches are in one tile, i.e. tile_size // patch_size.
+        num_layers (int): The number of transformer layers.
+        layer (nn.Module): The transformer layer module.
+        token_pos_embedding (nn.Module): The token positional embedding module.
+        pre_tile_pos_embed (Optional[nn.Module]): The pre-tile positional embedding module. It should be
+            None if your image was not tile-cropped in advance.
+        post_tile_pos_embed (Optional[nn.Module]): The post-tile positional embedding module. It should be
+            None if your image was not tile-cropped in advance.
+        cls_projection (Optional[nn.Module]): The CLS projection module. It should take an input tensor
+            of shape (bsz * n_tiles, n_tokens, embed_dim) and output a tensor of shape
+            (bsz * n_tiles, cls_output_dim). If provided, only the CLS token projection will be
+            outputted, instead of all tokens.
+        indices_return_hidden (Optional[List[int]]): The indices of hidden layers to return.
+            If provided, it will return the intermediate results of the transformer layers
+            before they go through a next layer. For example, indices_return_hidden = [0, 3] will
+            return the tokens before they go through the first and fourth layers.
+        patch_size (int): The size of each patch. Used to divide the tiles into patches.
```

Comment:
What is the difference between patch_grid_size and patch_size? They seem to both control the same parameter. Am I expected to set both of these?

Reply:
patch_grid_size is tile_size // patch_size. I could change it to tile_size instead, and do the calculation internally. The same would need to happen in the positional encodings. I was not sure about which one to pick.

Reply:
I guess that passing tile_size and calculating it internally is more intuitive, and more config friendly